library/src/main/java/com/novoda/downloadmanager/lib/BatchRepository.java

```diff
@@ -168,23 +163,39 @@ public DownloadBatch retrieveBatchBy(DownloadInfo downloadInfo) {
             batchesCursor.close();
         }
 
-        if (!forDeletion.isEmpty()) {
-            deleteBatchesForIds(forDeletion, downloads);
+        return batches;
+    }
+
+    public void deleteMarkedBatchesFor(Collection<DownloadInfo> downloads) {
+        Cursor batchesCursor = resolver.query(Downloads.Impl.BATCH_CONTENT_URI, null, SELECT_MARKED_FOR_DELETION, null, null);
```

Comment:
We should use a projection for performance though, we only need `_ID` from the rows.